sonar-scm-svn-plugin/src/main/java/org/sonar/plugins/scm/svn/FindFork.java

```diff
@@ -0,0 +1,111 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.scm.svn;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.Optional;
+import javax.annotation.CheckForNull;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.tmatesoft.svn.core.ISVNLogEntryHandler;
+import org.tmatesoft.svn.core.SVNException;
+import org.tmatesoft.svn.core.SVNLogEntry;
+import org.tmatesoft.svn.core.SVNLogEntryPath;
+import org.tmatesoft.svn.core.wc.SVNClientManager;
+import org.tmatesoft.svn.core.wc.SVNRevision;
+import org.tmatesoft.svn.core.wc.SVNStatus;
+
+import static org.sonar.plugins.scm.svn.SvnPlugin.newSvnClientManager;
+
+public class FindFork {
+  private static final Logger LOG = Loggers.get(FindFork.class);
+
+  private final SvnConfiguration configuration;
+
+  public FindFork(SvnConfiguration configuration) {
+    this.configuration = configuration;
+  }
+
+  @CheckForNull
+  public Instant findDate(Path location, String referenceBranch) throws SVNException {
+    ForkPoint forkPoint = find(location, referenceBranch);
+    if (forkPoint != null) {
+      return forkPoint.date();
+    }
+    return null;
+  }
+
+  @CheckForNull
+  public ForkPoint find(Path location, String referenceBranch) throws SVNException {
+    SVNClientManager clientManager = newSvnClientManager(configuration);
+    SVNRevision revision = getSvnRevision(location, clientManager);
+    LOG.debug("latest revision is " + revision);
+    String svnRefBranch = "/" + referenceBranch;
+
+    SVNLogEntryHolder handler = new SVNLogEntryHolder();
+    SVNRevision endRevision = SVNRevision.create(1);
+    SVNRevision startRevision = SVNRevision.create(revision.getNumber());
+
+    do {
+      clientManager.getLogClient().doLog(new File[] {location.toFile()}, startRevision, endRevision, true, true, -1, handler);
+      SVNLogEntry lastEntry = handler.getLastEntry();
+      Optional<SVNLogEntryPath> copyFromReference = lastEntry.getChangedPaths().values().stream()
+        .filter(e -> e.getCopyPath() != null && e.getCopyPath().equals(svnRefBranch))
+        .findFirst();
+
+      if (copyFromReference.isPresent()) {
+        return new ForkPoint(String.valueOf(copyFromReference.get().getCopyRevision()), Instant.ofEpochMilli(lastEntry.getDate().getTime()));
+      }
+
+      SVNLogEntryPath firstChangedPath = lastEntry.getChangedPaths().values().iterator().next();
+      if (firstChangedPath.getCopyPath() == null) {
+        // we walked the history to the root, and the last commit found had no copy reference. Must be the trunk, there is no fork point
+        return null;
+      }
+
+      // TODO should we iterate through other copy paths?
```

Comment:
Can you elaborate more on that?

Reply:
I wrote more in the comment.
Basically, based on the API, looks like a revision can have multiple changed paths. In other words, it can be a copy from multiple other branches. We only take the first one to keep looking for the fork. Imagine branch A which is forked from branches B and C, and C is forked from D. If we are in branch A and request it to find the fork with D, it won't find it because it will go through A -> B -> end.

Reply:
I don't know in what conditions it can happen so I didn't bother with it.